byron/semantics/executable-spec/src/Control/State/Transition/Generator.hs

```diff
@@ -122,11 +125,11 @@ data TraceLength = Maximum Word64 | Desired Word64
 
 data TraceProfile sts
   = TraceProfile
-  { proportionOfValidSignals :: Int
+  { proportionOfValidSignals :: !Int
     -- ^ Proportion of valid signals to generate.
-  , proportionOfInvalidSignals :: Int
+  , proportionOfInvalidSignals :: !Int
     -- ^ Proportion of invalid signals to generate.
-  , failures :: [(Int, PredicateFailure sts)]
+  , failures :: ![(Int, PredicateFailure sts)]
     -- ^ List of failure conditions to try generate when generating an invalid signal, and the
     -- proportion of each failure.
```

Comment:
We could remove the proportions of invalid signals, and use something like a ratio of valid signals. The reason why I used two proportions (valid/invalid) was because it aligned better with the `Gen.frequency` combinator, but we don't have to sacrifice clarity because of this. 

How would you propose to change the `TraceProfile`? Something like:

```haskell
data TraceProfile sts
  = TraceProfile
  { invalidSignalsRatio :: Ratio
  , failures :: ... as before
  }
```